### PR TITLE
ERB template rendering code review: rename old bosh v1 names to v2

### PR DESCRIPTION
**Pull request by bgandon**
### What is this change about?

Here, I’ve reviewed the ERB template rendering code in order to document it in the cloudfoundry/docs-bosh#813 PR.

In that part of the Director’s code, I’ve noticed _many_ misleading use of the term “template”, used instead of “job”. And this makes the code very confusing because “template” could also refer to “ERB template”.

So, I’ve started chasing these names that had not been migrated back in 2017 (when Bosh v1→v2 refactoring effort had been conducted), and I’ve looked for more suitable names. Don’t hesitate to challenge these, by the way.

Then I’ve updated the unit tests. Here also I’ve tried to clarify some ambiguous namings.

In that process, I’ve noticed that when rendering ERB templates, the job name was read from `job.MF`, freshly extracted from the job blob, whereas the name already persisted in database. I’ve removed the unnecessary unmarshalling of `job.MF` and added a name verification at release-upload time.

The modifications in this PR are quite large overall, but I’ve paid great attention in separating this into smaller chunks of work in to the individual commits.

### Please provide contextual information.

This work was done in order to gather information for authoring cloudfoundry/docs-bosh#813.

### What tests have you run against this PR?

Successful run of these test suites:

```shell
bundle exec rake spec:unit:parallel

DB=sqlite     bundle exec rake spec:unit:director
DB=postgresql bundle exec rake spec:unit:director
DB=mysql      bundle exec rake spec:unit:director
```

### How should this change be described in bosh release notes?

This change brings no new feature, and only more clarity in the ERB rendering code.

### Does this PR introduce a breaking change?

Definitely not.

### Tag your pair, your PM, and/or team!

I’m a [poor lonesome cowboy](https://www.pipelinecomics.com/wp-content/uploads/2020/10/lucky_luke_75_the_end.jpg?ezimgfmt=ngcb1/notWebP) 😄

### CODE REVIEW

File: src/bosh-director-core/lib/bosh/director/core/templates/job_instance_renderer.rb
Comment (aramprice) [inline src/bosh-director-core/lib/bosh/director/core/templates/job_instance_renderer.rb:54]: Should this be `spec_object['job']['name']` instead of `spec_object['name']`?

File: src/bosh-director/lib/bosh/director/deployment_plan/instance_group.rb
Comment (aramprice) [inline src/bosh-director/lib/bosh/director/deployment_plan/instance_group.rb:412]: agregated => aggregated

File: src/bosh-director/lib/bosh/director/deployment_plan/instance_group.rb
Comment (aramprice) [inline src/bosh-director/lib/bosh/director/deployment_plan/instance_group.rb:420]: agregated => aggregated

File: src/bosh-director/lib/bosh/director/jobs/update_release.rb
Comment (aramprice) [inline src/bosh-director/lib/bosh/director/jobs/update_release.rb:309]: At some point maybe the "#{name}/#{version}" moniker should be extracted into a method rather than having this pattern of string interpolation.

I don't think this is necessary for this PR but I thought it was worth including a comment in case there are more refactors coming.

File: src/bosh-director-core/spec/bosh/director/core/templates/job_template_loader_spec.rb
Comment (aramprice) [inline src/bosh-director-core/spec/bosh/director/core/templates/job_template_loader_spec.rb:82]: Is there a reason for the change from `release` => `plop`? If this is a meaningless identifier maybe use something like `FAKE_XYZ...`

File: src/bosh-director-core/spec/bosh/director/core/templates/job_template_loader_spec.rb
Comment (aramprice) [inline src/bosh-director-core/spec/bosh/director/core/templates/job_template_loader_spec.rb:117]: Is there a reason for the change from `plan` => `plop`? If this is a meaningless identifier maybe use something like `FAKE_XYZ...` or `FAKE_123...`

File: src/bosh-template/lib/bosh/template/evaluation_context.rb
Comment (aramprice) [inline src/bosh-template/lib/bosh/template/evaluation_context.rb:43]: Curious about this change, is this always safe, even if a v1 manifest is being consumed?

**Comment by rkoster** [comment]
@bgandon did you get a change yet to look at the changes requested by Aram?

File: src/bosh-director-core/lib/bosh/director/core/templates/job_instance_renderer.rb
Comment (bgandon) [reply src/bosh-director-core/lib/bosh/director/core/templates/job_instance_renderer.rb:54]: I’ve deliberately switched to using `spec.name` instead of `spec.job.name`, which holds the same information anyway.

For the reason why, first see: https://bosh.io/docs/jobs/#deprecated-properties-accessors

Indeed, I’ve documented there the `spec.job.*` properties (where “job” is the Bosh v1 name for “intance group” and “ template” the old Bosh v1 name for “job”, so that’s old). To my understanding, these have accidentally leaked in the `spec` object, as there is no reason for `spec.job.*` to be exposed to release authors because the data there are not useful for rendering ERB templates.

That’s why I’ve started favoring `spec.name` over `spec.job.name`. For preparing future removal of `spec.job`.

File: src/bosh-director/lib/bosh/director/jobs/update_release.rb
Comment (bgandon) [reply src/bosh-director/lib/bosh/director/jobs/update_release.rb:257]: About `Models::Template` representing a “job”, I’ve updated the comment of the class there:
https://github.com/cloudfoundry/bosh/pull/2479/files#diff-56d25d21526860acc6209368812592a1819f9db61fd20db419dabdca49173433R2-R5

I’ve thought this to be sufficient for the moment. Tell me what you think about this.

For a first refactoring, I’ve also deliberately chosen not to rename `Models::Template` in order to avoid renaming tables, and broader and more risky impacts. But it would be good to do it at some point.

File: src/bosh-director/lib/bosh/director/jobs/update_release.rb
Comment (bgandon) [reply src/bosh-director/lib/bosh/director/jobs/update_release.rb:309]: Yep, I agree with this indeed

**Comment by bgandon** [comment]
I’m in the process of reviewing the comments

File: src/bosh-director-core/spec/bosh/director/core/templates/job_template_loader_spec.rb
Comment (bgandon) [reply src/bosh-director-core/spec/bosh/director/core/templates/job_template_loader_spec.rb:82]: Yep, “plop” just conventionally mean “_whatever name_” and is meant to be more fun to use than “fake”. “plip” is quite the same, maning “_whatever name, but different from ‘plop’_”.

All right for favoring “fake” over “plip/plop” 👍

File: src/bosh-director-core/spec/bosh/director/core/templates/job_template_loader_spec.rb
Comment (bgandon) [reply src/bosh-director-core/spec/bosh/director/core/templates/job_template_loader_spec.rb:134]: There has been a reason indeed 🙂. Having different names for `plan-job-name` and `release-job-name` was a defect, that hadn’t show up before because `spec.name` wasn’t used.

There were no reason for distinguishing “plan” or “release” job, because there where designating the very same thing. These two names actually had to be the same. That’s how I started to use “plop“, and then when I needed a different name later on, I’ve used “plip”.

But I get your point. I’ll use `fake-job-name-1` and `fake-job-name-2` instead.

File: src/bosh-director-core/spec/bosh/director/core/templates/job_template_loader_spec.rb
Comment (bgandon) [reply src/bosh-director-core/spec/bosh/director/core/templates/job_template_loader_spec.rb:117]: I’ll use `fake-job-name-1` and `fake-job-name-2` instead of plip/plop.

File: src/bosh-template/lib/bosh/template/evaluation_context.rb
Comment (bgandon) [reply src/bosh-template/lib/bosh/template/evaluation_context.rb:43]: Considering that “job” means “instance group”, if the instance group is not a Hash then what could it be? An array? `nil`? Is this something that can happen?

To my knowledge, from within an ERB file that needs being rendered, the instance group _must_ be a Hash. If it’s not, then something bad is happening and then the code should crash.

Basically I’ve applied the “_let it crash_” principle: no reason to protect against a situation that should never happen, and possibly maintain unnecessary code or hide a bug.

File: src/bosh-director-core/spec/bosh/director/core/templates/job_template_loader_spec.rb
Comment (aramprice) [reply src/bosh-director-core/spec/bosh/director/core/templates/job_template_loader_spec.rb]: I kinda figured this was the case. I think over time we've found that  "fun" names tend to be less helpful that boring descriptive names. Sadly.

**Review by aramprice — APPROVED** [review]
thanks for the improvements!

**Comment by bgandon** [comment]
Due to the large diff, we are waiting for a second review, thanks for anyone that can do it!

**Review by jpalermo — APPROVED** [review]
👍

A lot of changes, but functionally seems to be the same. A lot of great cleanup though.

Thanks!

**Comment by rkoster** [comment]
Thanks! @bgandon

**Comment by aramprice** [comment]
This change appears to have exposed a problem with [src/spec/assets/valid_release.tgz](https://github.com/cloudfoundry/bosh/blob/main/src/spec/assets/valid_release.tgz).

Integration tests which attempt to deploy this release using a command similar to:
> `bosh [redacted args] upload-release src/spec/assets/valid_release.tgz`

Fail with the following error:
```
Using environment 'https://127.0.0.1:62404' as client 'test'
[-----------------------------------------------------------------] 100.00%   0s
Task 6

Task 6 | 17:40:22 | Extracting release: Extracting release (00:00:00)
Task 6 | 17:40:22 | Verifying manifest: Verifying manifest (00:00:00)
Task 6 | 17:40:22 | Resolving package dependencies: Resolving package dependencies (00:00:00)
Task 6 | 17:40:22 | Creating new packages: stuff/0.1.17 (00:00:01)
Task 6 | 17:40:23 | Creating new packages: mutator/2.99.7 (00:00:00)
Task 6 | 17:40:23 | Creating new jobs: cacher/1a (00:00:00)
                 L Error: Inconsistent name for job 'cacher'(exptected: 'cacher', got: '')
Task 6 | 17:40:23 | Error: Inconsistent name for job 'cacher'(exptected: 'cacher', got: '')

Task 6 Started  Thu Jan 11 17:40:22 UTC 2024
Task 6 Finished Thu Jan 11 17:40:23 UTC 2024
Task 6 Duration 00:00:01
Task 6 error
```

This appears to be run from [spec/integration/release/export_release_spec.rb:222](https://github.com/cloudfoundry/bosh/blob/main/src/spec/integration/release/export_release_spec.rb#L222)

**Comment by beyhan** [comment]
I didn't have time to get to the bottom of this but it looks like there is a new validation now https://github.com/cloudfoundry/bosh/blob/dcb90acf53b19bee64ac07ed0565e29fb5e1b847/src/bosh-director/lib/bosh/director/jobs/release/release_job.rb#L79-L84

Actually this change shouldn't have any behaviour changes. @bgandon why that check was added?

**Comment by aramprice** [comment]
@beyhan - nice catch! No need to dig in I only figured this out at the end of my yesterday.

It's likely that `src/spec/assets/valid_release.tgz` is making some "bosh v1" assumptions.

**Comment by aramprice** [comment]
Hopefully fixed with https://github.com/cloudfoundry/bosh/commit/e42f7670484516ad0dcfb2ae8375fb051db085b2

**Comment by beyhan** [comment]
@aramprice thanks for the heads-up. I also saw that the last update of that release was way back in 2015 and had an old structure. Is the support for "bosh v1" removed with this refactoring or it was already the case and the validation here just shows it? If it is with this pr I would suggest to have this in the release notes. FYI @rkoster, @jpalermo

**Comment by jpalermo** [comment]
I think v1 manifests should still work, I don't think release job spec file structure changed much between v1 and v2, and I would have thought it was "required" that jobs in v1 also had a name that matched their actual name.

I kinda bet something blows up at some point if this isn't true, so I don't think this is really a change in behavior, probably just surfaces the error sooner. (could be wrong about all of this though)

**Pull request merged**
